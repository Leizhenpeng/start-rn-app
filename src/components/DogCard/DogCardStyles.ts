import {StyleSheet} from 'react-native'

export const styles = StyleSheet.create({
  card: {
    flex: 1,
    margin: 10,
    backgroundColor: '#fff',
    borderRadius: 8,
    overflow: 'hidden',
    borderColor: '#ddd',
    borderWidth: 1,
  },
  image: {
    width: '100%',
    height: 150, // Adjust height based on your needs
    borderTopLeftRadius: 8,
    borderTopRightRadius: 8,
  },
  info: {
    padding: 10,
  },
  name: {
    fontSize: 16,
    fontWeight: 'bold',
  },
  bredFor: {
    fontSize: 14,
    color: '#666',
  },
})

