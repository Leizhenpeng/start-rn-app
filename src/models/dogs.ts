export interface Weight {
  imperial: string // 重量（英制）
  metric: string // 重量（公制）
}

export interface Height {
  imperial: string // 身高（英制）
  metric: string // 身高（公制）
}

export interface Breed {
  weight: Weight // 重量
  height: Height // 身高
  id: number // 品种ID
  name: string // 品种名称
  bred_for: string // 培育用途
  breed_group: string // 品种组
  life_span: string // 寿命
  temperament: string // 性情
  reference_image_id: string // 参考图片ID
  country_code?: string // 国家代码（可选）
}

export interface Dog {
  breeds: Breed[] // 品种信息数组
  id: string // 图片ID
  url: string // 图片URL
  width: number // 图片宽度
  height: number // 图片高度
}
