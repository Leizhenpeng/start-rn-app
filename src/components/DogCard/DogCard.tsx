import React from 'react'
import {Image, Text, View} from 'react-native'
import {styles} from './DogCardStyles'
import {Dog} from '~/models/dogs'
import FastImage from 'react-native-fast-image'

type DogCardProps = {
  item: Dog
}

const DogCard: React.FC<DogCardProps> = ({item}) => {
  const breed = item.breeds[0]
  const imageUrl = item.url

  return (
    <View style={styles.card}>
      <FastImage source={{uri: imageUrl}} style={styles.image} />
      {breed && (
        <View style={styles.info}>
          <Text style={styles.name}>{breed.name}</Text>
          <Text style={styles.bredFor} numberOfLines={2}>{breed.bred_for}</Text>
        </View>
      )}
    </View>
  )
}

export default DogCard

