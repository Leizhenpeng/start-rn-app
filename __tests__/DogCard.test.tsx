import React from 'react'
import {render} from '@testing-library/react-native'
import DogCard from '~/components/DogCard'
import {Dog} from '~/models/dogs'

// Mock data for testing
const mockDog: Dog = {
  breeds: [
    {
      weight: {
        imperial: '40 - 70',
        metric: '18 - 32',
      },
      height: {
        imperial: '17 - 20',
        metric: '43 - 51',
      },
      id: 81,
      name: 'Chow Chow',
      bred_for: 'Guardian, cart pulling, hunting',
      breed_group: 'Non-Sporting',
      life_span: '12 - 15 years',
      temperament: 'Aloof, Loyal, Independent, Quiet',
      reference_image_id: 'ry8KWgqEQ',
    },
  ],
  id: 'ry8KWgqEQ',
  url: 'https://cdn2.thedogapi.com/images/ry8KWgqEQ_1280.jpg',
  width: 1010,
  height: 673,
}

describe('DogCard', () => {
  it('renders correctly with given dog data', () => {
    const {getByText, getByTestId, toJSON} = render(<DogCard item={mockDog} />)

    // Check if the breed name and bred_for are displayed
    expect(getByText('Chow Chow')).toBeTruthy()
    expect(getByText('Guardian, cart pulling, hunting')).toBeTruthy()

    // Optionally, snapshot testing to ensure component output remains consistent
    expect(toJSON()).toMatchSnapshot()
  })
})
