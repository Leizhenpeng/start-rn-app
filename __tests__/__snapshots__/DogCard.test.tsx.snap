// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DogCard renders correctly with given dog data 1`] = `
<View
  style={
    {
      "backgroundColor": "#fff",
      "borderRadius": 8,
      "flex": 1,
      "margin": 10,
      "overflow": "hidden",
    }
  }
>
  <View
    style={
      [
        {
          "overflow": "hidden",
        },
        {
          "borderTopLeftRadius": 8,
          "borderTopRightRadius": 8,
          "height": 150,
          "width": "100%",
        },
      ]
    }
  >
    <FastImageView
      defaultSource={null}
      resizeMode="cover"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
  </View>
  <View
    style={
      {
        "padding": 10,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      Chow Chow
    </Text>
    <Text
      numberOfLines={2}
      style={
        {
          "color": "#666",
          "fontSize": 14,
        }
      }
    >
      Guardian, cart pulling, hunting
    </Text>
  </View>
</View>
`;
